subsys/bluetooth/mesh/proxy.c

```diff
@@ -956,6 +956,7 @@ bool bt_mesh_proxy_relay(struct net_buf *buf, uint16_t dst)
 		bt_mesh_adv_send_start(0, err, BT_MESH_ADV(buf));
 		if (err) {
 			BT_ERR("Failed to send proxy message (err %d)", err);
+			net_buf_unref(buf);
```

Comment:
Should this `unref` not go inside `proxy_segment_and_send`, as it's that function's reference?

Reply:
Yes, seems you're right. There's otherwise at least one other place that's missing an unref() after calling proxy_segment_and_send()

Reply:
Hold on, does proxy_segment_and_send() take ownership of the reference then? If so it seems this code is not allowed to use buf in the call to bt_mesh_adv_send_start() anymore.

Reply:
https://github.com/zephyrproject-rtos/zephyr/blob/f6d7f8f36fffdf2b311252eee54f765d854a51b7/subsys/bluetooth/mesh/proxy.c#L953-L954

Reply:
@LingaoM thanks, I missed that. Then it means `proxy_segment_and_send()` will consume (unref or pass onward) the buffer no matter what.

Reply:
> Then it means proxy_segment_and_send() will consume (unref or pass onward) the buffer no matter what.

This is not implemented as far as I can tell. The unref here is an unref of the ref we pass to segment_and_send, which that function should handle itself. If not, then this unref should be unconditional, no?

Reply:
@trond-snekvik Do you mean unref inside this `proxy_segment_and_send` ?

Reply:
https://github.com/zephyrproject-rtos/zephyr/blob/8aeac35c5d7cf83d3d2a4d77b8bef5d72f9a10ea/subsys/bluetooth/mesh/proxy.c#L163-L166
The last params type `void *`.

Reply:
I see, that makes it a bit awkward. Seems wrong to unref the reference we just passed to that function though. If anything, I suppose it's this function that should get a ref in its parameters, and not segment_and_send. 

This appears to be working correctly, so I'll approve. @jhedberg are you happy with this unref?

Reply:
Yes, it seems it's correct. It wouldn't hurt to have a code comment however, something along the lines of "If segment_and_send() fails the buf_send_end() callback will not be called, so we need to clear the user data (net_buf) reference given to segment_and_send() here"